### PR TITLE
fix(ds): Fix unsubscribe logic related to the replay

### PR DESCRIPTION
**Pull request by ieQu1**
Fixes https://emqx.atlassian.net/browse/EMQX-11537

Release version: v/e5.?

## Summary

Postpone deletion of UNSUBSCRIBE until all in-flight packets are processed by the client. 

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx/src/emqx_persistent_session_ds.erl [hunk 1]
```diff
@@ -313,11 +313,12 @@ subscribe(
...
                 props => SubOpts,
-                id => SubId
+                id => SubId,
+                deleted => false
             },
             IsNew = true;
         Subscription0 = #{} ->
-            Subscription = Subscription0#{props => SubOpts},
+            Subscription = Subscription0#{props => SubOpts, deleted => false},
             IsNew = false,
```
Comment (ieQu1) [inline apps/emqx/src/emqx_persistent_session_ds.erl:322]: Take `deleted` into account

File: apps/emqx/src/emqx_persistent_session_ds.erl [hunk 2]
```diff
@@ -336,15 +337,22 @@ unsubscribe(
...
             {ok, Session#{s => S}, SubOpts}
     end.
 
 -spec do_unsubscribe(id(), topic_filter(), subscription(), emqx_persistent_session_ds_state:t()) ->
     emqx_persistent_session_ds_state:t().
-do_unsubscribe(SessionId, TopicFilter, #{id := SubId}, S0) ->
-    S1 = emqx_persistent_session_ds_state:del_subscription(TopicFilter, [], S0),
+do_unsubscribe(SessionId, TopicFilter, SubMeta0 = #{id := SubId}, S0) ->
+    %% Note: we cannot delete the subscription immediately, since its
+    %% metadata can be during replay (see `process_batch'). We instead
```
Comment (ieQu1) [inline apps/emqx/src/emqx_persistent_session_ds.erl:351]:
```suggestion
    %% metadata can be used during replay (see `process_batch'). We instead
```

File: apps/emqx/src/emqx_persistent_session_ds_stream_scheduler.erl [hunk 3]
```diff
@@ -303,6 +357,12 @@ compare_streams(
 is_fully_replayed(Comm1, Comm2, S = #srs{it_end = It}) ->
     It =:= end_of_stream andalso is_fully_acked(Comm1, Comm2, S).
 
+is_fully_acked(_, _, #srs{
+    first_seqno_qos1 = Q1, last_seqno_qos1 = Q1, first_seqno_qos2 = Q2, last_seqno_qos2 = Q2
```
Comment (thalesmg) [inline apps/emqx/src/emqx_persistent_session_ds_stream_scheduler.erl:361]: Q: a freshly created `#srs{}` will be in this condition; should it be considered fully acked as well?

Comment (ieQu1) [reply apps/emqx/src/emqx_persistent_session_ds_stream_scheduler.erl:361]: Yes, I don't see why we can't fetch from a freshly discovered stream immediately.

**Pull request merged**